internal/k8s/watch.go

```diff
@@ -0,0 +1,134 @@
+package k8s
+
+import (
+	"context"
+	"log"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// PodScrapeDetails stores the metrics endpoint details and metadata for a pod.
+type PodScrapeDetails struct {
+	Port      string
+	Path      string
+	PodName   string
+	Namespace string
+}
+
+// PodScrapeWatcher manages pod metrics and provides methods to handle updates and deletions.
+type PodScrapeWatcher struct {
+	PodMetricsEndpoints map[string]PodScrapeDetails
+	mu                  sync.Mutex
+
+	// Function variables for update and delete operations, to allow mocking during tests.
+	UpdatePodMetricsFunc func(*corev1.Pod)
+	DeletePodMetricsFunc func(*corev1.Pod)
+	HandlePodEventFunc   func(watch.Event)
+}
+
+// NewPodScrapeWatcher initializes a new PodScrapeWatcher with default function implementations.
+func NewPodScrapeWatcher() *PodScrapeWatcher {
+	pw := &PodScrapeWatcher{
+		PodMetricsEndpoints: make(map[string]PodScrapeDetails),
+	}
+	pw.HandlePodEventFunc = pw.HandlePodEvent
+	pw.UpdatePodMetricsFunc = pw.UpdatePodMetrics
+	pw.DeletePodMetricsFunc = pw.DeletePodMetrics
+
+	return pw
+}
+
+// GetPodMetricsEndpoints returns the current pod metrics endpoints.
+func (pw *PodScrapeWatcher) GetPodMetricsEndpoints() map[string]PodScrapeDetails {
+	pw.mu.Lock()
+	defer pw.mu.Unlock()
+
+	// Return a copy of the PodMetricsEndpoints to avoid race conditions
+	endpointsCopy := make(map[string]PodScrapeDetails)
+	for k, v := range pw.PodMetricsEndpoints {
+		endpointsCopy[k] = v
+	}
+	return endpointsCopy
+}
+
+// WatchPods watches for pod changes and updates the metrics endpoints accordingly.
+func (pw *PodScrapeWatcher) WatchPods(clientset kubernetes.Interface, namespace string, labels map[string]string) {
+	labelSelector := metav1.FormatLabelSelector(&metav1.LabelSelector{MatchLabels: labels})
+	for {
+		watcher, err := clientset.CoreV1().Pods(namespace).Watch(context.TODO(), metav1.ListOptions{
+			LabelSelector: labelSelector,
+		})
+		if err != nil {
+			log.Fatalf("Error watching pods: %v", err)
+		}
+
+		for event := range watcher.ResultChan() {
+			pw.HandlePodEventFunc(event)
+		}
```

Comment:
Do we care about graceful shutdowns?  I think probably not, but raising it just in case. 
 If we do, this would be an issue as it'll block indefinitely.  We'd need a context or something else in here if we want to support ending early.

Reply:
This will be tackled in a separate issue.